polkadot/runtime/parachains/src/paras_inherent/mod.rs

```diff
@@ -617,8 +618,24 @@ impl<T: Config> Pallet<T> {
 
 				(candidate_receipt_with_backing_validator_indices, backed_candidates_with_core)
 			} else {
+				log::debug!(
+					target: LOG_TARGET,
+					"Upcoming session change, not backing any new candidates."
+				);
 				// If we'll initialize a new session at the end of the block, we don't want to
 				// advance the claim queue.
+
+				// In `Enter` context (invoked during execution) no more candidates should be
+				// filtered, because they have already been filtered during `ProvideInherent`
+				// context. Abort in such cases.
+				// In the case of an upcoming session change, the backed candidates should be empty.
+				if context == ProcessInherentDataContext::Enter {
+					ensure!(
+						initial_candidate_count == 0,
+						Error::<T>::CandidatesFilteredDuringExecution
```

Comment:
As mentioned elsewhere (not introduced here): These checks should really be more top-level.